### ISSUE TITLE
SR point-wise evaluation with measuring "stitcher" performance 

### ISSUE THREAD

**Issue by keighrim** [clamsproject/aapb-evaluations#60]
### New Feature Summary

#55 added an evaluation software for apps like SWT only using `TimePoint` annotations, but that evaluation can easily expanded to evaluation of "stitcher" component that turns `TP` annotations into `TimeFrame` annotations. The idea is based on that all existing stitcher implementation is using "label remapping", exposed as a runtime param and recorded in the view metadata, that enables us to re-construct point-wise but remapped label value list. 

So the idea is to update `eval.py` file so that it 

1. Reads `TP` annotation as usual, constructing a list of "raw" classification results, let's call it `raw`
2. Reads the gold files as usual, constructing a list of gold labels, let's call it `gold`
1. Grabs the "latest" view with `TimeFrame` annotation with the remapper config (`map` for SWT built-in, `labelMap` for simple-stitcher)
1. Uses the remapper to map the `raw` and `gold` into secondary remapped lists (these new lists should be shorter then the original ones, since not all of the raw/gold labels are remapped into the secondary (`TF`) labels. Let's call them `raw-remap` and `gold-remap` respectively. 
1. Then iterates through the `TF` annotations, construct a third list of stitched, remapped labels by using the pointers in `targets` prop (which must be pointing to `TP` annotations so the timepoints can be traced), let's call this list `stitched`
1. compute P/R/F between 
    - `raw` vs. `gold` (this should already be there in the current eval.py) 
    - `raw-remap` vs. `gold-remap`
    - `stitched` vs. `gold-remap`, 

### Related

resolving this issue will also properly address https://github.com/clamsproject/app-swt-detection/issues/61

**Comment by marcverhagen** [comment]
Some of this was done in 22bee5c (work in progress that I just pushed up to make it visible), but not quite in the way as described above. One difference is that it uses the output of the updated [process.py](https://github.com/clamsproject/aapb-annotations/blob/73f6d308a147cb84447d7dee529b06bb9de4be4f/scene-recognition/process.py) script from the annotations repository.

**Comment by marcverhagen** [comment]
Is there overlap here with https://github.com/clamsproject/aapb-evaluations/issues/43?

**Comment by keighrim** [comment]
The evaluation scheme in this is based on point-wise evaluation, hence is not compatibly with "old" interval-level gold data (from 2020 time). I don't think this is a duplicate of #43. Eventually, I believe the proposed method will be evaluation for stitcher components, largely independent from image classification model performance. 

I think the most "overlapping" effort to this issue was `evaluate.py`[^1] in https://github.com/clamsproject/app-swt-detection/commit/5590a3e975afa9008687a8533aa167bebb35dc9f  , but that file was 
1. never used to produce a public report or associated with any actual (archived) SWT MMIF output files
2. written a fair while ago and the current status/compatibility is unknown

so I thought it's not easy to verify if the old code works or not. (plus, this repo is the repo for evaluation code) That led me to start this new issue. 

[^1]: which wasn't easy to dig out, since the most closely related issue - presumably https://github.com/clamsproject/app-swt-detection/issues/61 - was closed without mentioning the file or commit or an explicit PR merge.

**Comment by keighrim** [comment]
For future references, the "result" files used for this gridsearch/evaluation; 
[swt6.1-stitcher3.0-results.zip](https://github.com/user-attachments/files/16749471/swt6.1-stitcher3.0-results.zip)

**Comment by keighrim** [comment]
A few follow up questions; 

- how would stitching without binning (remap) perform? (i.e., just using the single letter lables)
- why does 6way remap work poorly in general? 
- 5 secs seems to be too long threshold for chyrons, what happened to chyron? 

And future directions;
- merge stitcher into SWT, using smoother implementation from SDK
  - this is based on an assumption that this particular implementation of stitching only work for SWT (or likes) app (other apps, for example https://github.com/clamsproject/app-brandeis-acs-wrapper, might use a completely different algorithm and runtime parameters)
- make SWT works in three modes - TP-only, TF-only (stitcher mode), and TP+TF mode 
  - separate stitcher can add a better modularization, but since SWT is the only app that uses this stitching implementation, the benefit can't outweigh the user-friendliness of having a single app with all related features built-in.

**Comment by kla7** [comment]
> how would stitching without binning (remap) perform? (i.e., just using the single letter lables)

I'm unsure what you mean by this since the current stitching mechanics involve remapping TPs into TFs.

> 5 secs seems to be too long threshold for chyrons, what happened to chyron?

I don't remember why I did not specifically report on `chyron` or `slate`, but both seemed to show great results all around. I will do further reports below to expand on the previous findings and highlight `chyron` and `slate`.

> why does 6way remap work poorly in general?

My assumption is that some timeframes predicted to be `other_text` or `other_opening` should actually be labeled as `chyron`, `credits`, or `slate` and vice versa. I will include visualizations to observe 6way vs. 4way remap while normalizing `minTFDuration = 1000` (to better support `chyrons` by tackling the threshold concern mentioned above) and `minTFScore = 0.5` (which I have found previously to produce the best results).

For `chyron` and `slate`, the F/P/R scores tend to be around 0.02 points lower using 6way remap compared to 4way remap.

Sample comparison between the two label map presets for `chyron` to demonstrate these findings:
![image](https://github.com/user-attachments/assets/0a7120a8-a2f0-4a8d-bace-fd1eb8b3a4de)

Sample comparison between the two label map presets for `slate` to demonstrate these findings:
![image](https://github.com/user-attachments/assets/1c0ef1b5-3c5a-44e2-a9e7-78e749b6980d)

For `credits`, R scores are minimally affected (<0.01 difference), but P scores are between 0.15-0.04 points lower using 6way remap compared to 4way remap, while F1 scores are around 0.03-0.04 points lower. From these numbers, I suspect that many timeframes that should not be labeled as `credits` is predicted as such, judging by the near perfect R scores in contrast with lower P scores (supported by lower F1 scores).

Sample comparison between the two label map presets for `credits` to demonstrate these findings:
![image](https://github.com/user-attachments/assets/7b591770-dc87-4668-9c78-64f4be3cde01)

I am now acknowledging my prior misunderstanding of how `minTFDuration` affects the results, since setting it to 5000ms can lead to the loss of valuable data as not every TF lasts 5000ms. When `minTFDuration` is set to 1000ms, the difference between the `filtered` and `stitched` F/P/R scores is still significant.

When observing results where `minTFDuration = 1000` and `minTFScore = 0.5`, it is interesting to see that F/P/R scores across the three labels (again, `bars` has been excluded since it scores perfectly across all configs) are slightly higher when `minTPScore = 0.001` compared to when `minTPScore = 0.5`. I can't figure out a reason for why this might be the case. To my understanding, `minTPScore` determines which TPs are included so I seem to be missing an intuitive piece here.

**Comment by keighrim** [comment]
> How would stitching without binning (remap) perform?

### Experiment setup
The goal of this experiment is not to find the best set of parameters that fit the optimized binning scheme, instead the goal is to find a "good-enough" set of values that works fairly well for "generally complex" label spaces. 
#### Input data
- aapb-collaboration-27-[abcd] (62 videos)
- all of these 62 videos are used as training set for the CV model for timepoint annotations, hence we should expect the accuracy of the CV model on these videos would be "better" than usual, hence we won't look at the absolute measurement of the model accuracy. We'll discuss the evaluation method used in this experiment later in this report. 
#### TimePoint annotations
- (relevant) configurations
```
        "app": "http://apps.clams.ai/swt-detection/v6.1",        
        "appConfiguration": {
          "sampleRate": 1000, 
          "modelName": "convnext_lg",
          "usePosModel": true, 
        }                                                                                                     
```
#### TimeFrame annotations
- (relevant) configurations
```

        "appConfiguration": {
          "useStitcher": true,
          "tfMinTPScore": OR(0.001, 0.055, 0.5),
          "tfMinTFScore": OR(0.055, 0.5),
          "tfMinNegTFDuration": OR(1, 500, 1000, 2000),
          "tfMinTFDuration": OR(2000, 3000, 5000),
          "tfAllowOverlap": OR(True, False)
        }

```
- in the above, values in `OR()` notation is used for "gridsearch" in the experiment
- why 0.055 in score thresholds? that's roughly 1/18 (even distribution of all labels)
- `tfMinNegTFDuration` parameter is only used for this experimental setup, and won't be exposed in the release version. This parameter was proposed in https://brandeis-llc.slack.com/archives/C06B0B3TYVB/p1727397593356179 
#### "binning"
- no binning was use in TP (prebin)
    - at TP level, we have total 18 (`BSWLOMINEPYKUGTFCR`) "raw" labels (ignoring subtypes of slates)
    - in 6.1, by mistake `U` was omitted from the model
    - as a result, `BSWLOMINEPYKGTFCR-` (raw - `U` + `-`) were used in the TP view of the experimental data
- no binning was used in TF (postbin) annotation 
    - namely, the labels for time frames were also the (buggy) "raw" single-letter labels. 
- some "bin names" used in the following are simply meant for averaging scores 

### Results
#### Evaluation method
- Based on discussion in https://github.com/clamsproject/aapb-evaluations/issues/60 (and implementation in https://github.com/clamsproject/aapb-evaluations/pull/62), we calculated point-wise precision/recall/F-1 scores for "unfiltered" and "stitched" (since we didn't use any binning/mapping, "unfiltered" and "filtered" conditions are identical, and hence ignored)
- To measure the contribution of the stitcher, we subtract unfiltered P/R/F from stitched P/R/F scores. In other words, we consider the differences in point-wise P/R/F scores as the contribution (larger position difference = contributing larger improvement, vice versa for negative)
- We then visualized the results with their associated parameter values as [PC](https://en.wikipedia.org/wiki/Parallel_coordinates) using [HiPlot](https://github.com/facebookresearch/hiplot) python library. 
#### result files
- html files (link)
    - visualization files are named as `stitcher-results-{AVERAGE_BIN}-{[PRF]}-diff.html`
    - `AVBERAGE_BIN` is NOT pre- or postbinning used in the SWT/stitcher app. These bins are used just in the visualization to compute average P/R/F scores of certain groups of raw (single-letter) labels


#### Analysis
##### When looking at the all labels 
<details>
<summary>label vs all scores (diffs)</summary>

![Pasted image 20241101131826](https://github.com/user-attachments/assets/614a23b9-ace6-404d-ae8e-b0487c8eacc2)
- `W` label showed the widest variance in the grid search, from top to bottom. 
</details>
<details>
<summary>label vs positive scores (the top one improvement of `W` is intentionally excluded)</summary>

![Pasted image 20241101132113](https://github.com/user-attachments/assets/64764ccf-78bd-4b29-b3a9-21f7fb0381c9)
- most of the higher improvements (red colored lines) are observed in `T`, `O`, `G`, `F` labels, which are not of our immediate interest 
</details>
<details>
<summary>point-wise softmax threshold vs score</summary>


![Pasted image 20241101132512](https://github.com/user-attachments/assets/e3069cf3-8127-41b1-8120-c9d0c1cc4106)
- extremely low threshold (0.001) that was used the previous iteration showed the widest impact on the stitcher contribution, including lots of negative ones. 
- the middle value (even distribution, or "by-chance" baseline) showed mostly positive effect. 

![Pasted image 20241101133805](https://github.com/user-attachments/assets/45c838ad-6ec1-4051-ba45-fb3ab0839a56)
- the true majority threshold (0.5) always showed improvements.

![Pasted image 20241101140536](https://github.com/user-attachments/assets/837a02a8-0dd0-46bd-956c-9086f6b59682)
- when the threshold is high, `minNegTFDuration` (used to "smooth-up" short noisy misclassification in the middle of longer sequences) values didn't make meaningful differences. 

![Pasted image 20241101140616](https://github.com/user-attachments/assets/c26f77af-1f64-4f37-b6f7-c4558807dfc0)
![Pasted image 20241101140724](https://github.com/user-attachments/assets/6310370e-d8a8-496b-9197-339a24948956)
- the same goes for `minTFScore` and `minTFDuration`. 
</details>

###### conclusion: using `minTPScore=0.5` will make a generally better-working stitcher. 

##### When looking at labels in "relaxed" binning scheme
> [!NOTE]: again, this is not binning of model, but binning used to compute the average P/R/F scores. 
<details>
<summary>label vs score</summary>

![Pasted image 20241101135905](https://github.com/user-attachments/assets/a7d4757f-010a-426c-a0d4-9f7ec5574670)
- similar to the above (all the raw labels)
</details>
<details>
<summary>average of all "interested" labels vs score</summary>

![Pasted image 20241101140142](https://github.com/user-attachments/assets/d4887d74-f3b3-495b-b59e-cb89e33b0a6d)
![Pasted image 20241101140209](https://github.com/user-attachments/assets/14ecb0fc-0daf-49df-a93f-8a9c7c3e14ff)
- The top ranked instances showed ~16% F-1 score improvements in average across all the "interesting" labels. 

![Pasted image 20241101140318](https://github.com/user-attachments/assets/a20b4834-4249-4028-9545-6dbfa3dca9b5)
- and we can re-affirm that `minTPScore=0.5` resulted in positive impact all the time. 
</details>
<details>
<summary> by-bin vs score</summary>

![Pasted image 20241101141409](https://github.com/user-attachments/assets/04ff30aa-77b9-4c94-acc9-9c4be0c4f70f)
![Pasted image 20241101141431](https://github.com/user-attachments/assets/fedb6fd9-955a-464c-9439-8dabc7478da6)
![Pasted image 20241101141453](https://github.com/user-attachments/assets/a48462ac-fc2d-4414-aea6-8fec2672acf9)
![Pasted image 20241101141511](https://github.com/user-attachments/assets/227c43da-1afe-4c83-a3a4-919d91a145ea)
![Pasted image 20241101141526](https://github.com/user-attachments/assets/56cc04c0-d0c3-4cb0-90f2-2defd32fdf2d)
- when the softmax threshold is fixed at 0.5, "other-text" from the relaxed scheme showed the largest improvement (~20%, or 0.18 - 0.22) via the stitcher, while "chyron" and "credits" showed ~10% difference, and for "slate" and "bars" the stitcher showed only limited improvements. 

![image](https://github.com/user-attachments/assets/44e345a6-420d-4263-87b7-055785cb6f26)
![image](https://github.com/user-attachments/assets/f852ba42-def9-4bdd-adf2-5d210241ff13)
- interestingly, in a previous discussion, we suspected 5-second `minTFDuration` threshold can be too harsh for chyron types, but when `minTPScore` is high, the duration 5-sec duration threshold showed the largest improvement, primarily due to its positive impact on `N` and `Y` labels. 
</details>

###### conclusion: as shown in the previous pilot report by @kla7, `minTFDuration=5000` will make a generally better-working stitcher (with combination with `minTPScore=0.5`)

### Conclusion and next steps
I will fix `tfMinTPScore=0.5` and `tfMinTFDuration=5000` parameters and will conduct a similar experiment with the "relaxed" postbin applied at the stitching time.

**Comment by keighrim** [comment]
# 241102 experiment with "relaxed" postbin
## Experiment setup

This is an experiment following up the findings from `241101` experiment on stitcher contribution without any binning. The goal of this experiment is to find a set of proper defaults values for the remainder of the stitcher configuration that works well for relatively simpler label spaces after a "post-binning" of single-letter raw labels. 
### Input data
- The same TP annotations on 62 videos ([`aapb-collaboration-27-[abcd]`](https://github.com/clamsproject/aapb-annotations/tree/24d202a8857baadeeda31172c954f11d4111cde6/batches)) were used as input to the stitcher. 
### TimeFrame annotations
- (relevant) configurations
```

        "appConfiguration": {
          "useStitcher": true,
          "tfMinTPScore": 0.5,
          "tfMinTFScore": OR(0.5, 0.75, 0.9),
          "tfLabelMapFn": OR("sum", "max"),
          "tfMinNegTFDuration": OR(1, 1000, 2000),
          "tfMinTFDuration": 5000,
          "tfAllowOverlap": OR(True, False)
        }

```
- In the above, values in `OR()` notation is used for "gridsearch" in the experiment. Total number of parameter combinations is 3 x 2 x 3 x 2 = 36. 
- A new experimental parameter is added in this round; `tfLabelMapFn`. This is the function to use to aggregate a point-wise confidence score when binning two or more sublabels. In the previous versions of SWT, this function is fixed to `max`, meaning given a binning from [`I`, `N`] to `chyron`, the score of the `chyron`, or `S(chyron)` is determined by the maximum number from [`S(I)`, `S(N)`]. In this experiment, I tried switching it to `sum` function (i.e., `S(chyron)` = `S(I)` + `S(N)`) to see if that makes any difference. 
- As we found from the last `241101` experiment, I fixed `minTPScore` and `minTFDuration` values to 0.5 and 5-second, respectively. 
### "binning"
- no binning for TP (prebin) - still the same (buggy) raw labels were used as the output space of the SR model. 
- "relaxed" postbin (`"Overall-relaxed"` in [here](https://github.com/clamsproject/app-swt-detection/issues/118#issuecomment-2408152019)) applied on time point annotations before they are stitched into time frames. 

## Results
### Evaluation method
- I used essentially the same method from the previous experiment, except that now we applied postbinning, the P/R/F scores of "unfiltered" (before postbinning) and "filtered" (after postbinning) are different. To measure the contribution of the stitching step that only happens after postbinning, I use the difference between "stitched" P/R/F and "filtered" as the measurement. 
- Then I used the same visualization method ([parallel coordinates](https://en.wikipedia.org/wiki/Parallel_coordinates) using [HiPlot](https://github.com/facebookresearch/hiplot) python library). 
### Result files
- html files ([stitcher-results-241102.zip](https://github.com/user-attachments/files/17608974/stitcher-results-241102.zip))

    - visualization files are named as `stitcher-results-!-{[PRF]}-{CONDITION}.html`
    - `CONDITION` is one of `filtered`, `stitched`, or `diff` that means point-wise accuracy of "filtered" labels, "stitched" labels, and the difference ("stitched" - "filtered"), respectively. 

## Analysis 

#### Stitcher contribution by all bins
- With all parameter combinations, I couldn't see any one of them negatively impacting the final accuracy. Particularly, "credits" showed a wider range of stitcher contribution, followed by "slate" and "other-text". "chyron" and "bars" seemed already quite stable before stitching. 
    <details>
    <summary>visuals</summary>

    ![Screenshot_20241102_211948](https://github.com/user-attachments/assets/3039f36a-a191-476a-a209-f38b5ccf3fbc)

    </details>
- As seen in the "filtered" scores, "bars", "slate", and "chyron" all reached pretty high accuracy before stitching (as expected by the "leak" of the training data). 
    <details>
    <summary>visuals</summary>

    ![Screenshot_20241102_212256](https://github.com/user-attachments/assets/a1fa44c4-f44f-480c-b0c7-f4e6f4e80664)


    </details>

#### Softmax aggregation method vs stitcher contribution
- No significant difference is found between `max` and `sum` functions. This is probably because we are already using a true majority (0.5) as the softmax threshold. 
    <details>
    <summary>visuals</summary>

    ![Screenshot_20241102_212434](https://github.com/user-attachments/assets/166ba0ce-3c15-4027-99f6-ac03cc43d96f)
    ![Screenshot_20241102_212437](https://github.com/user-attachments/assets/859c3cba-26ef-438b-823c-cc3ef3bb5d19)

    </details>

#### Frame-wise score (average) threshold
- Since we use `minTPScore=0.5`, I experimented with some values greater than or equal to that (0.5, 0.75, 0.9) for `minTFScore` parameter. 
- Overall, the higher threshold, the more positively the stitcher contributed. 
    <details>
    <summary>visuals</summary>

    ![Screenshot_20241102_212955](https://github.com/user-attachments/assets/2457e178-b136-4195-9b67-7b6e6446b622)
    ![Screenshot_20241102_212615](https://github.com/user-attachments/assets/fad99d31-3805-4c85-bfec-c57aa755a0cb)

    </details>
- 0.9 frame-wise score threshold boost both precision and recall more than 0.75. 
    <details>
    <summary>visuals</summary>

    ![Screenshot_20241102_220601](https://github.com/user-attachments/assets/a16ee6e2-2116-4e53-b317-80db2b8cb59a)
    ![Screenshot_20241102_220534](https://github.com/user-attachments/assets/de56d42f-931d-42bd-aeee-9de9dee6e7e7)
    ![Screenshot_20241102_220612](https://github.com/user-attachments/assets/3aade565-96b8-48e7-90c3-e058a9640f16)
    </details>
- Especially for "credits", and less so for "slate" and "other-text".
    <details>
    <summary>visuals</summary>

    ![Screenshot_20241102_213050](https://github.com/user-attachments/assets/3fec19c1-a113-4a9d-ad98-2b53425d4c19)
    ![Screenshot_20241102_213048](https://github.com/user-attachments/assets/44d52928-c161-43d8-866c-c17ea44587cf)
    ![Screenshot_20241102_213046](https://github.com/user-attachments/assets/4e929481-837e-4a5b-b6fc-f7b332a9aea7)
    ![Screenshot_20241102_213044](https://github.com/user-attachments/assets/48a99467-f465-4709-870b-9cfcc7a654b2)
    ![Screenshot_20241102_213042](https://github.com/user-attachments/assets/6f940374-8971-4523-b168-11729e63a927)


    </details>

#### Smoothing negative "noises" 
- `minTFNegTFDuration` parameter is used to filter out short negative noises. 
- Note that in the previous stitcher implementation (SWT v6.1 or older), this "negative smoothing" wasn't implemented, and in the new implementation, `minNegTFDuration` values less than or equal to the timepoint sampling rate becomes equivalent to the old implementation, or lack thereof, with `minTFNegDuration=1`. being absolute "kill switch" for the negative smoothing feature.
- Since in this experiment, the sampling rate of the timepoints are 1000 ms, I simply ignored `minTFNegDuration=1` and compared 1000 and 2000 ms.
- Between 1- vs. 2-second thresholds, looking at average of all bins, 2-sec showed a wider range of positive contribution, sometimes more, sometimes less than 1-sec threshold.
    <details>
    <summary>visuals</summary>

    ![Screenshot_20241102_213719](https://github.com/user-attachments/assets/d56068e2-8109-4247-b3cb-ef2603015835)

    </details>
- But when combined with `minTFScore` parameter, clearly the longer negative smoothing threshold benefits from the higher frame-wise score threshold, especially for "credits", without hurting the accuracy on other bins.
    <details>
    <summary>visuals</summary>

    [Screencast_20241102_214818.webm](https://github.com/user-attachments/assets/ebc6e8c2-6212-43cb-9aac-45b9c4df36c6)
    </details>


#### Lastly, allowing frame overlap
- When `minTFScore` and `minTFNegDuration` parameters are fixed to the best configuration based on the findings so far (0.9 and 2-sec, respectively), `allowOverlap` parameter didn't show any differences. 
    <details>
    <summary>visuals</summary>

    ![Screenshot_20241102_174057](https://github.com/user-attachments/assets/f48989ac-036a-407d-8c91-3edc118bf528)
    </details>
- This is probably because of the high frame-wise score threshold, which makes the overlapping time frames less likely to happen.
- Since overlapping time frames will pose significant complexity to downstream apps, I recommend using `allowOverlap=False` as the default value.

### Conclusion

- In general, stitching works as intended for the smaller label space after postbinning. 
- For the two "experimental" parameters, 
    - `labelMapFn` is likely shadowed by the hight `minTPScore=0.5` threshold. I recommend keeping it as `max` for consistency with the previous versions.
    - `minTFNegDuration` (negative smoothing) alone showed a mixed results (compared to turning it off), but when combined with higher `minTFScore` threshold, it showed a clear accuracy boost. I recommend hardcoding 2000 ms threshold, and not exposing it to the user to reduce the complexity of the app. 
- Finally, for the default values for the stitcher parameters, I recommend using `minTFScore=0.9`, `allowOverlap=False`.
